[Clang][SYCL] Introduce clang-sycl-link-wrapper to link SYCL offloading device code (#1)
asudarsa/llvm-project: clang/docs/ClangSYCLLinkWrapper.rst

> ====================
> Clang SYCL link Wrapper
> ====================

Reviewer: ```suggestion
=======================
Clang SYCL link Wrapper
=======================
```

.. contents::
   :local:

.. _clang-sycl-link-wrapper:

Introduction
============

This tool works as a wrapper around the SYCL device code linking process.

> Purpose of this wrapper is to provide an interface to link SYCL device bitcode

Reviewer 2: ```suggestion
The purpose of this wrapper is to provide an interface to link SYCL device bitcode
```

> Purpose of this wrapper is to provide an interface to link SYCL device bitcode

Reviewer: According to my understanding, the purpose of the tool is to provide standard linker interface and be able to link any object file formats 'compile' step is able to produce for SYCL offload mode. So, I would expect this tool to handle linking of fat objects with device code in other IR like SPIR-V as well as native binary format.

Author: Hi @bader 

Thanks for the feedback. I agree with you. This tool is expected to handle linking of fat objects and SPIR-V objects. However, in its current state, it can support only bitcode files. We can extend this tool easily to support linking of other objects. I was planning to update the document as the tool evolves. I will add more details to the doc.

Reviewer: Thanks! Let's clearly state the final goal for the tool, current status of the implementation, and plans for the future work.

in LLVM IR format, run some SYCL-specific finalization steps and then use the
SPIR-V LLVM Translator tool to produce the final output.

> Device code linking for SYCL offloading kind has a number of known quirks that

Reviewer 2: ```suggestion
Device code linking for SYCL offloading has a number of known quirks that
```

makes it difficult to use in a unified offloading setting. Two of the primary
issues are:
1. Several finalization steps are required to be run on the fully-linked LLVM

> IR bitcode to gaurantee conformance to SYCL standards. This step is unique to

Reviewer 2: ```suggestion
IR bitcode to guarantee conformance to SYCL standards. This step is unique to
```

> SYCL offloading compilation flow.

Reviewer 2: ```suggestion
the SYCL offloading compilation flow.
```

> 2. SPIR-V LLVM Translator tool is an extenal tool and hence SPIR-V IR code

Reviewer 2: ```suggestion
2. SPIR-V LLVM Translator tool is an external tool and hence SPIR-V IR code
```

generation cannot be done as part of LTO. This limitation will be lifted once

> SPIR-V backend is available as a viable LLVM backend.

Reviewer 2: ```suggestion
the SPIR-V backend is available as a viable LLVM backend.
```

This tool works around these issues.

Usage
=====

This tool can be used with the following options. Several of these options will

> be passed down to downstrea tools like 'llvm-link', 'llvm-spirv', etc.

Reviewer 2: ```suggestion
be passed down to downstream tools like 'llvm-link', 'llvm-spirv', etc.
```

Author: Will run spellcheck on this doc.

> be passed down to downstrea tools like 'llvm-link', 'llvm-spirv', etc.

Reviewer: ```suggestion
be passed down to downstrea tools like 'clang', 'llvm-spirv', etc.
```
We should use `clang` instead of `llvm-link` to link llvm bitcode files.

Author: I can add a TODO. I do have a tracker to replace llvm-link with clang. '-only-needed' linking support seems to be missing in clang.

Reviewer: I think `llvm-link` is not a part of clang distribution, so SYCL linking won't work. We need to resolve that ASAP.

.. code-block:: console

  OVERVIEW: A utility that wraps around the SYCl device code linking process.

>   This enables linking and code generation for SPIR-V JIT targets.

Reviewer: What about AOT targets?

Author: I was planning to add it as the implementation evolves in near future. But I will add it now and provide a PR comment about upcoming changes.

Thanks

  USAGE: clang-sycl-link-wrapper [options]

  OPTIONS:
    --arch <value>       Specify the name of the target architecture.
    --dry-run            Print generated commands without running.
    -g                   Specify that this was a debug compile.
    -help-hidden         Display all available options
    -help                Display available options (--help-hidden for more)
    --library-path=<dir> Set the library path for SYCL device libraries
    -o <path>            Path to file to write output
    --save-temps         Save intermediate results
    --triple <value>     Specify the target triple.
    --version            Display the version number and exit
    -v                   Print verbose information

Example
=======

This tool is intended to be invoked when targeting the SPIR-V toolchain.

> When --sycl-link option is passed, clang driver will invoke the linking job of

Reviewer 2: ```suggestion
When the `--sycl-link option` is passed, the clang driver will invoke the linking job of
```

> When --sycl-link option is passed, clang driver will invoke the linking job of

Reviewer: Please, specify that `--sycl-link` is the `clang` driver option, not `clang-sycl-link-wrapper` option.

the SPIR-V toolchain, which in turn will invoke this tool.

> This tool can be used to create one or more fully linked SYCL objects that are

Reviewer: ```suggestion
This tool can be used to create one or more fully linked device images that are
```
?

=======================
Clang SYCL Link Wrapper
=======================

.. contents::
   :local:

.. _clang-sycl-link-wrapper:

Introduction
============

This tool works as a wrapper around the SYCL device code linking process.
The purpose of this wrapper is to provide an interface to link SYCL device
bitcode in LLVM IR format, SYCL device bitcode in SPIR-V IR format, and native
binary objects, and then use the SPIR-V LLVM Translator tool on fully linked
device objects to produce the final output.
After the linking stage, the fully linked device code in LLVM IR format may
undergo several SYCL-specific finalization steps before the SPIR-V code
generation step.

> The wrapper will also support the Ahead-Of-Time compilation flow. AOT

Reviewer 3: ```suggestion
The wrapper will also support the Ahead-Of-Time (AOT) compilation flow. AOT
```

compilation is the process of invoking the back-end at compile time to produce
the final binary, as opposed to just-in-time (JIT) compilation when final code

> generation is deferred until application runtime time.

Reviewer 3: ```suggestion
generation is deferred until application runtime.
```

Device code linking for SYCL offloading has several known quirks that

> makes it difficult to use in a unified offloading setting. Two of the primary

Reviewer 3: ```suggestion
make it difficult to use in a unified offloading setting. Two of the primary
```

Author: 'makes' goes well with 'Device code linking'. WDYT?

Reviewer 3: It reads interesting, as the association can also be with 'quirks'.  It is the quirks that make it difficult - it's not the device code linking that makes it difficult.

be passed down to downstream tools like 'llvm-link', 'llvm-spirv', etc.

.. code-block:: console

  OVERVIEW: A utility that wraps around the SYCL device code linking process.
  This enables linking and code generation for SPIR-V JIT targets and AOT
  targets.

  USAGE: clang-sycl-link-wrapper [options]

  OPTIONS:
    --arch <value>                Specify the name of the target architecture.
    --dry-run                     Print generated commands without running.
    -g                            Specify that this was a debug compile.
    -help-hidden                  Display all available options
    -help                         Display available options (--help-hidden for more)
    --library-path=<dir>          Set the library path for SYCL device libraries
    -o <path>                     Path to file to write output
    --save-temps                  Save intermediate results
    --triple <value>              Specify the target triple.
    --version                     Display the version number and exit
    -v                            Print verbose information
    -spirv-dump-device-code=<dir> Directory to dump SPIR-V IR code into

>     -sycl-is-windows-msvc-env     Specify if we are compiling under windows environment

Reviewer 3: Do we need 'sycl' here?  The tool is a SYCL specific tool - seems redundant.
